### ISSUE TITLE
wrapper to the hypre eigensolver

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mfem/mfem#2]
```
Adding a wrapper to the hypre eigensolver and a corresponding modification of 
ex3/ex3p, please? 

Andrew Knyazev
```

Original issue reported on code.google.com by `2AndrewK...@gmail.com` on 9 Aug 2012 at 9:56

**Comment by martemyev** [comment]
Hi all,
I'd like to vote for this wrapper as I need an eigensolver for a sparse generalized eigenproblem for which LOBPCG is applicable.

Best regards,
Mikhail

**Comment by mlstowell** [comment]
We have just finished a first draft of an MFEM wrapper for the HYPRE_LOBPCG eigenvalue solver and a modification of ex1p on which to test it.  We can certainly build another test based on ex3/ex3p but the larger null space of that operator concerns me.  The wrapper still needs to pass a design review and I expect the interface to need a few minor adjustments before it's ready for release.

These changes will appear in the next release of MFEM which is expected to be available by the end of the year.

Thanks for your interest!

Mark

**Comment by martemyev** [comment]
Hi Mark,

Actually, I'm not as interested in ex3p as Andrew, so I'd be really grateful if you can show the first draft of the wrapper and the ex1p using it. I was trying to do that myself, but still couldn't make it work, so even if the design is not perfect it would be very nice to see how it can be done in principle.

Best regards,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

Here is an outline of one possible interface for wrapping _hypre_'s LOBPCG. Note that this integration is still under active development, and the code in the mfem release will probably look (significantly) different.

Hope this is still helpful for you,
Tzanio

```c++
class HypreMultiVector
{
private:
   mv_MultiVectorPtr mv_ptr;
   mv_InterfaceInterpreter interpreter;
   HypreParVector ** hpv;
   int nv;

public:
   HypreMultiVector(int n, HypreParVector & v);
   ~HypreMultiVector();

   void Randomize(HYPRE_Int seed) { mv_MultiVectorSetRandom(mv_ptr, seed); }
   HypreParVector & GetVector(unsigned int i) { return *hpv[i]; }
   operator mv_MultiVectorPtr() const { return mv_ptr; }
   mv_InterfaceInterpreter & GetInterpreter() { return interpreter; }
   mv_MultiVectorPtr       & GetMultiVector() { return mv_ptr; }
};

class HypreLOBPCG
{
private:
   HYPRE_Solver lobpcg_solver;
   HYPRE_MatvecFunctions matvec_fn;

public:
   HypreLOBPCG(mv_InterfaceInterpreter & interpreter);
   ~HypreLOBPCG();

   void SetTol(double tol)
   {  HYPRE_LOBPCGSetTol(lobpcg_solver, tol); }
   void SetMaxIter(int max_iter)
   { HYPRE_LOBPCGSetMaxIter(lobpcg_solver, max_iter);}
   void SetPrintLevel(int logging)
   { HYPRE_LOBPCGSetPrintLevel(lobpcg_solver, logging); }
   void SetPrecondUsageMode(int pcg_mode)
   { HYPRE_LOBPCGSetPrecondUsageMode(lobpcg_solver, pcg_mode); }

   void SetPrecond(HypreSolver & precond);
   void Setup(HypreParMatrix & A, HypreParVector & b, HypreParVector & x);
   void SetupB(HypreParMatrix & B, HypreParVector & x);
   void SetupT(HypreParMatrix & T, HypreParVector & x);

   void Solve(Vector & eigenvalues)
   { HYPRE_LOBPCGSolve(lobpcg_solver, NULL, NULL, eigenvalues); }
   void Solve(Vector & eigenvalues, HypreMultiVector & eigenvectors)
   { HYPRE_LOBPCGSolve(lobpcg_solver, NULL, eigenvectors, eigenvalues); }
};

HypreMultiVector::HypreMultiVector(int n, HypreParVector & v) : nv(n)
{
   HYPRE_ParCSRSetupInterpreter(&interpreter);
   mv_ptr = mv_MultiVectorCreateFromSampleVector(&interpreter, nv,
                                                 (HYPRE_ParVector)v);
   HYPRE_ParVector* vecs = NULL;
   {
      mv_TempMultiVector* tmp =
         (mv_TempMultiVector*)mv_MultiVectorGetData(mv_ptr);
      vecs = (HYPRE_ParVector*)(tmp -> vector);
   }
   hpv = new HypreParVector*[nv];
   for (int i=0; i<nv; i++)
   {
      hpv[i] = new HypreParVector(vecs[i]);
   }
}

HypreMultiVector::~HypreMultiVector()
{
   for (int i=0; i<nv; i++)
   {
      if ( hpv[i] != NULL ) { delete hpv[i]; }
   }
   if ( hpv != NULL ) { delete [] hpv; }

   mv_MultiVectorDestroy(mv_ptr);
}

HypreLOBPCG::HypreLOBPCG(mv_InterfaceInterpreter & interpreter)
{
   HYPRE_ParCSRSetupMatvec(&matvec_fn);
   HYPRE_LOBPCGCreate(&interpreter, &matvec_fn, &lobpcg_solver);
}

HypreLOBPCG::~HypreLOBPCG()
{
   HYPRE_LOBPCGDestroy(lobpcg_solver);
}

void HypreLOBPCG::SetPrecond(HypreSolver & precond)
{
   HYPRE_Solver solver = (HYPRE_Solver)precond;
   HYPRE_LOBPCGSetPrecond(lobpcg_solver,
                          (HYPRE_PtrToSolverFcn)precond.SolveFcn(),
                          (HYPRE_PtrToSolverFcn)precond.SetupFcn(),
                          solver);
}

void HypreLOBPCG::Setup(HypreParMatrix & A, HypreParVector & b, HypreParVector & x)
{
   HYPRE_ParCSRMatrix parcsr_A = A;
   HYPRE_LOBPCGSetup(lobpcg_solver,(HYPRE_Matrix)parcsr_A,
                     (HYPRE_Vector)((HYPRE_ParVector)b),
                     (HYPRE_Vector)((HYPRE_ParVector)x));
}

void HypreLOBPCG::SetupB(HypreParMatrix & B, HypreParVector & x)
{
   HYPRE_ParCSRMatrix parcsr_B = B;
   HYPRE_LOBPCGSetupB(lobpcg_solver,(HYPRE_Matrix)parcsr_B,
                      (HYPRE_Vector)((HYPRE_ParVector)x));
}

void HypreLOBPCG::SetupT(HypreParMatrix & T, HypreParVector & x)
{
   HYPRE_ParCSRMatrix parcsr_T = T;
   HYPRE_LOBPCGSetupT(lobpcg_solver,
                      (HYPRE_Matrix)parcsr_T,
                      (HYPRE_Vector)((HYPRE_ParVector)x));
}
```

**Comment by martemyev** [comment]
Hi Tzanio,
Thanks a lot! It's very helpful.
Could you please explain a bit about the Setup function? What are the vectors 'b' and 'x' for?

Thank you,
Mikhail

**Comment by tzanio** [comment]
You can ignore them, they are passed but not really used. Historically they come from _hypre_'s multigrid solvers, which have Setup() and Solve() functions with the same signatures.

**Comment by martemyev** [comment]
Tzanio,
I was able to compile and run the eigensolver, so thanks for the wrapper.
There is a question, though - how can I describe if I want the eigenvalues (and corresponding eigenvectors) with the smallest or largest magnitude?

Thank you,
Mikhail

**Comment by martemyev** [comment]
I'm also not sure if I implemented everything correctly. Here is my code:
```cxxx
int main(int argc, char *argv[])
{
   int num_procs, myid;
   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myid);

   int nx = 100, ny = 100;
   int order = 1;

   Mesh *mesh = new Mesh(nx, ny, Element::QUADRILATERAL, 1);
   int dim = mesh->Dimension();

   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
   delete mesh;

   FiniteElementCollection *fec = new H1_FECollection(order, dim);
   ParFiniteElementSpace *fespace = new ParFiniteElementSpace(pmesh, fec);
   int size = fespace->GlobalTrueVSize();
   if (myid == 0)
      cout << "Number of unknowns: " << size << endl;

   ConstantCoefficient one(1.0);

   ParBilinearForm *a = new ParBilinearForm(fespace);
   a->AddDomainIntegrator(new DiffusionIntegrator(one));
   a->Assemble();
   a->Finalize();
   HypreParMatrix *S = a->ParallelAssemble();
   if (num_procs == 1)
   {
     ofstream out("matrix.dat");
     S->PrintMatlab(out);
   }

   HypreParVector b(*S), x(*S), v(*S);

   const int blockSize = 10;
   HypreMultiVector eigenvectors(blockSize, v);
   Vector eigenvalues(blockSize);

   HypreSolver *precond = new HypreBoomerAMG(*S); // AMG is used in the Hypre example ex11

   HypreLOBPCG hypre_lobpcg(eigenvectors.GetInterpreter());

   hypre_lobpcg.SetPrecond(*precond);
   hypre_lobpcg.Setup(*S, b, x);
   hypre_lobpcg.Solve(eigenvalues, eigenvectors);

   {
     char id[2];
     sprintf(id, "%d", myid);
     string fname = (string)"eigenvalues_" + id + ".dat";
     ofstream out(fname.c_str());
     eigenvalues.Print_HYPRE(out);
   }

   delete precond;

   delete S;
   delete a;

   delete fespace;
   delete fec;
   delete pmesh;

   MPI_Finalize();

   return 0;
}
```
When I compute the eigenvalues in Matlab, I get this:
```
>> load matrix.dat; M = spconvert(matrix);
>> eigs(M, 10, 'lm')
ans =
    3.9987
    3.9987
    3.9977
    3.9977
    3.9961
    3.9961
    3.9958
    3.9958
    3.9948
    3.9948
>> eigs(M, 10, 'sm')
ans =
    0.0086
    0.0076
    0.0047
    0.0047
    0.0038
    0.0038
    0.0019
    0.0010
    0.0010
    0.0000
```
The results of the c++ code shown above is:
```
$ cat eigenvalues_0.dat
10
6.89903156642557e-310
6.89903156642557e-310
0.00000000000000e+00
0.00000000000000e+00
7.65199591727200e-72
1.48269714088793e-76
1.03291347896428e-47
1.26931701069948e-76
8.97336404303815e-67
1.35887243077571e-153
```
So, I'm wondering what I'm doing wrong.

Thank you in advance,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

LOBPCG solves for the minimal eigenpairs of the generalized eigenvalue problem `A x = lambda B x`, where `A` is e.g. your stiffness matrix, and `B` is e.g. a corresponding mass matrix, or the identity. You need to also set `B` in your code by calling `HypreLOBPCG::SetupB`.

As far as I know, computing maximal (or interior) eigenvalues with _hypre_'s LOBPCG is not straightforward.

Hope this helps,
Tzanio

**Comment by martemyev** [comment]
Hypre code (version 2.10.1, file HYPRE_lobpcg.h) says that the B matrix is optional. Here is an extract:
```c
/**
 * (Optional) Set up {\tt B}.  If not called, B = I.
 **/
HYPRE_Int HYPRE_LOBPCGSetupB(HYPRE_Solver solver, 
                       HYPRE_Matrix B,
                       HYPRE_Vector x);
```
So, I guess that is okay to miss the B matrix. What else should I look at?

Thank you,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

You need an initial guess for the eigenvector subspace that is "rich enough", so call 

```c++
int seed = 42;
eigenvectors.Randomize(seed);
```
before `hypre_lobpcg.Solve`.

Also, the eigenvalue problem you are trying to solve, `Ax = lambda x` has a singular `A`, which LOBPCG does not handle well. Typically eigenvalue problems have prescribed zero boundary conditions, so the simplest thing to do is to add

```c++
Array<int> ess_bdr(pmesh->bdr_attributes.Max());
ess_bdr = 1;
a->EliminateEssentialBC(ess_bdr);
```
before `a->Finalize();`. You could also consider solving the equivalent eigenvalue problem `(A+I) x = (lambda+1) x` where the matrix is non-singular. 

The eigenvalues of the Laplacian (with zero boundary conditions) for the unit square on continuous level are of the form `(m^2+n^2) pi^2` where `m` and `n` are positive integers, so the minimal ones are:

```
2 pi^2  = 19.73920880217872
5 pi^2  = 49.34802200544679 (double)
8 pi^2  = 78.95683520871487
10 pi^2 = 98.69604401089359 (double)
13 pi^2 = 128.3048572141617 (double)
17 pi^2 = 167.7832748185191 (double)
18 pi^2 = 177.6528792196085
...
```
Since you don't have the mass matrix on the right-hand side, yours should be scaled additionally by about h^2.

Hope this helps,
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,
Thank you for your help with the eigensolver!
I have, however, a question about randomizing the initial guess. Without it the code clearly doesn't work, but I noticed that with different values of `int seed` I get slightly different eigenvalues. I'm solving a generalized eigenproblem S*x=lambda*M*x, where S and M are stiffness and mass matrix for a simple elastic problem. The zero Dirichlet boundary conditions are imposed. Here is my working minimal example:
```cxx
#include "mfem.hpp"
#include "hypre_wrapper.hpp" // for HypreLOBPCG
#include <iostream>
#include <fstream>
#include <ctime>

using namespace std;
using namespace mfem;

int main(int argc, char **argv)
{
  MPI_Init(&argc, &argv);
  int myid;
  MPI_Comm_rank(MPI_COMM_WORLD, &myid);
  char id[2];
  sprintf(id, "%d", myid);

  Mesh *mesh = new Mesh(10, 10, Element::QUADRILATERAL, 1, 1000., 1000.);
  int dim = mesh->Dimension();
  ParMesh pmesh(MPI_COMM_WORLD, *mesh);
  delete mesh;
  int order = 1;
  FiniteElementCollection *fecoll = new H1_FECollection(order, dim);
  ParFiniteElementSpace fespace(&pmesh, fecoll, dim);

  const double rho = 2500.;
  const double vp  = 3500.;
  const double vs  = 2000.;
  ConstantCoefficient lambda(rho*(vp*vp-2.*vs*vs));
  ConstantCoefficient mu(rho*vs*vs);
  ConstantCoefficient rhocoef(rho);

  Array<int> ess_bdr(pmesh.bdr_attributes.Max());
  ess_bdr = 1;
  int d = 1; // put 0 on the diagonal when finalizing

  ParBilinearForm stif(&fespace);
  stif.AddDomainIntegrator(new ElasticityIntegrator(lambda, mu));
  stif.Assemble();
  stif.EliminateEssentialBC(ess_bdr, d);
  stif.Finalize();
  HypreParMatrix *S = stif.ParallelAssemble();

  ParBilinearForm mass(&fespace);
  mass.AddDomainIntegrator(new VectorMassIntegrator(rhocoef));
  mass.Assemble();
  stif.EliminateEssentialBC(ess_bdr, d);
  mass.Finalize();
  HypreParMatrix *M = mass.ParallelAssemble();

  {
    string fname = (string)"stif_mat_" + id + ".dat";
    ofstream out(fname.c_str());
    S->PrintMatlab(out);
  }
  {
    string fname = (string)"mass_mat_" + id + ".dat";
    ofstream out(fname.c_str());
    M->PrintMatlab(out);
  }

  HypreParVector b(*S), x(*S), v(*S);

  int neigs = 10; // number of eigenvalues
  HypreMultiVector eigenvectors(neigs, v);
  int seed = atoi(argv[1]); // to play with the seed
  eigenvectors.Randomize(seed);
  Vector eigenvalues(neigs);

  HypreBoomerAMG *Sprec = new HypreBoomerAMG(*S);
  Sprec->SetPrintLevel(0);

  HypreLOBPCG hypre_lobpcg(eigenvectors.GetInterpreter());
  hypre_lobpcg.SetPrecondUsageMode(0);
  hypre_lobpcg.SetPrintLevel(0);

  hypre_lobpcg.SetPrecond(*Sprec);
  hypre_lobpcg.Setup (*S, b, x);
  hypre_lobpcg.SetupB(*M, x);

  hypre_lobpcg.Solve(eigenvalues, eigenvectors);

  {
    string fname = (string)"eigenvalues_" + id + ".dat";
    ofstream out(fname.c_str());
    eigenvalues.Print_HYPRE(out);
  }

  delete Sprec;
  delete fecoll;

  MPI_Finalize();
  return 0;
}
```
I assume the eigenvalues computed in Matlab are correct:
```
load ../build/stif_mat_0.dat;
S = spconvert(stif_mat_0);
load ../build/mass_mat_0.dat;
M = spconvert(mass_mat_0);
format longE
eigs(S, M, 10, 'sm')
     1.531466357473621e+02
     1.531466357473622e+02
     2.132195240874575e+02
     3.273610222904044e+02
     4.285262461739243e+02
     4.285262461739247e+02
     4.305940523728271e+02
     5.653865753021381e+02
     6.389659543000053e+02
     6.389659543000053e+02
```
Here are some results with LOBPCG (I run with 1 process) and different seed values:
```
./test 1; cat eigevalues_0.dat         <-- THIS IS OKAY
10
1.53146157552152e+02
1.53146157552155e+02
2.13219033569744e+02
3.27360529779488e+02
4.28525743797733e+02
4.28525743797736e+02
4.30593587858996e+02
5.65386115613323e+02
6.38965471749945e+02
6.38965471749960e+02
```
```
./test 42; cat eigevalues_0.dat
10
1.53146141333337e+02
1.53146157552149e+02
2.13219033317862e+02
3.27360519717044e+02
3.65126524171561e+02 <-- THIS ONE UNEXPECTED
4.28525626587404e+02
4.28525743797722e+02
4.30593518425655e+02
5.65386097706010e+02
6.38940007781973e+02
```
```
./test 450; cat eigevalues_0.dat
10
4.65504200439102e+00 <-- THIS ONE UNEXPECTED
1.53146157552141e+02
1.53146157552157e+02
2.13219033569748e+02
3.27360529779484e+02
4.28525743797740e+02
4.28525743797747e+02
4.30593587858991e+02
5.65386115613346e+02
6.38965471749497e+02
```
And so on. I would use seed=1, but it gives results different from Matlab for the 12x12 mesh, for example.
Matlab results are always in consistency with results that I get with LAPACK's dsygvd function for the same matrices.
So, the question is: What can I do to get more consistent eigenvalues?

Thank you in advance!

Best regards,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

The issue might be with the boundary conditions. Try eliminating them in both `stif` and `mass` putting a nonzero entry, say `1/eps`, on the diagonal in `stif` and an entry of `1.0` on the diagonal of `mass`  (you will need a new version of `EliminateEssentialBC` for that). This should shift the eigenvalues coming from the bc elimination to `1/eps`, so they shouldn't affect your computation for `eps` small enough.

Another issue could be the convergence of LOBPCG. What's the tolerance you are using?

Hope this helps,
Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,
Thanks for the ideas! I played with the tolerance, and already implemented a new version of `EliminateEssentialBC`, when I suddenly found out that I had the wrong code - I eliminated the essential bc twice for stiffness matrix, and not for the mass matrix (so you were right - that was indeed because of the boundary conditions). Shame on me. When I fixed that, everything works perfectly with all seeds I tried.

Thanks again!

Best regards,
Mikhail
